### PR TITLE
dsigm 0.3.1 staging

### LINKED ISSUE #2: SGMM.fit() stalls due to stabilization algorithm
**Describe the bug**

SGMM.fit() hangs when attempting self-stabilization

**To Reproduce**

```
SGMM().fit(some_data)
```

**Expected behavior**

SGMM.fit() should converge relatively fast.

**Desktop (please complete the following information):**
 - Python: 3.6.9
 - OS: OS-independent
 - Version: 0.3.0 prerelease

### LINKED ISSUE #4: SGMM with irregular data
**Ask your question**

When I run SGMM.fit() with no stabilization on some irregular data, it creates a very different result from sklearn's GMM. Not sure if this is acceptable?

Here's an example using `sklearn.datasets.make_moons`:
image

This is in contrast to sklearn's GMM:
image

### LINKED ISSUE #6: SGMM.fit() unstable results
**Describe the bug**

`SGMM.fit` results in different results each time.

**To Reproduce**

```
X_stretched = np.dot(X, np.random.RandomState(2).randn(2, 2))
sgmm = SGMM(init_cores=4, stabilize=None)
sgmm.fit(X_stretched)
```

**Expected behavior**

Should output same parameters each time.

**Screenshots**

Expected:
image

Occasionally gets this result:
image

**Desktop (please complete the following information):**
 - Python: 3.6.9
 - OS: OS-independent
 - Version: 0.3.1 prerelease

### PR DESCRIPTION
**Pull request by paradoxysm**
**What does this implement/fix? Explain your changes.**

This is a PR to merge `master` into `staging` for the 0.3.1 release.

**Additional comments**

Ensure that the project, [dsigm 0.3.1](https://github.com/paradoxysm/dsigm/projects/2), is completed before accepting this PR.

### CODE REVIEW

**Review by paradoxysm — COMMENTED** [review]
Tentative approval, review further tests.
Awaiting final guide on stabilization

**Review by paradoxysm — COMMENTED** [review]
Tentative approval!

File: doc/README.md [hunk 1]
```diff
@@ -8,4 +8,22 @@ All documentation for dsigm is located here!
...
+
+# SGMM
+**A Self-stabilizing Gaussian Mixture Model**
+
+The SGMM acts as the base clustering layer for the DSIGM algorithm. The SGMM is a modified Gaussian Mixture Model (GMM) capable of automatically determining an optimal number of components.
+
+When self-stabilization is disabled, the SGMM becomes a normal GMM, modelled after sklearn's GaussianMixture. Implementation will not be covered in depth here and can be found in sklearn's documentation [here](). In brief, the SGMM initializes *k* components using the k-means algorithm, and then using the Expectation-Maximization algorithm, iteratively updates the model parameters to maximize the log likelihood until convergence.
```
Comment (paradoxysm) [inline doc/README.md]: Link to sklearn.GaussianMixture is missing here

File: doc/README.md [hunk 2]
```diff
@@ -8,4 +8,22 @@ All documentation for dsigm is located here!
...
+
+Self-stabilization is based on an exploratory approach attempting to minimize the Bayesian Information Criterion (BIC). The BIC is a popular criterion for model selection with a strong penalty for excessive components to prevent overfitting. The BIC can be plotted over different number of components, *k*, to generate a curve where the minimum points to the ideal *k*, *k'*. Past usages of BIC for automatically finding *k'* typically initialize a model with one component and iteratively increment the number of components until the BIC has been minimized. While sufficient when *k'* is small, should *k'* be large, these methods are inefficient.
```
Comment (paradoxysm) [inline doc/README.md]: No longer minimizing on BIC, instead on composite ABIC

File: doc/pydoc/Core.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+## _validate_init
+```python
+Core._validate_init()
+```
+Validate the attributes of the Core.
+If `dim` has not yet been set, set it now.
+
+## pdf
+```python
+Core.pdf(data)
```
Comment (paradoxysm) [inline doc/pydoc/Core.md]: Add weighting argument

File: doc/pydoc/Core.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+**Returns**
+```
+pdf : ndarray or scalar
+	Probability density function evaluated at `data`.
+```
+
+## logpdf
+```python
+Core.logpdf(data)
```
Comment (paradoxysm) [inline doc/pydoc/Core.md]: Same as pdf, add weight argument

File: doc/pydoc/GMM.md [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+## score
+```python
+GMM.score(data)
+```
+Compute the per-sample average log-likelihood.
+
+**Parameters**
+```
+p_norm : array-like, shape (n_samples,)
```
Comment (paradoxysm) [inline doc/pydoc/GMM.md]: Reflect changes to using data

File: doc/pydoc/SGMM.md [hunk 6]
```diff
@@ -0,0 +1,395 @@
+# dsigm.SGMM
+
+The Self-stabilizing Gaussian Mixture Model (SGMM) is a modified Gaussian Mixture Model that is capable of automatically converging to the optimal number of components during fitting. The SGMM refines the number of components by narrowing a search interval through the Bayesian Information Criterion until converged.
```
Comment (paradoxysm) [inline doc/pydoc/SGMM.md]: No longer uses BIC, instead composite ABIC

File: doc/pydoc/SGMM.md [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+```
+interval : tuple, shape (2,)
+	The interval which contains the optimal number of Cores.
+	Interpreted as [min, max).
+
+bic : tuple, shape (2,)
+	The bic scores corresponding to the interval.
+```
+
+## _estimate_parameters
```
Comment (paradoxysm) [inline doc/pydoc/SGMM.md]: Remove the following as they belong in [GMM.md](https://github.com/paradoxysm/dsigm/tree/master/doc/pydoc/GMM.md)

File: dsigm/_core.py [hunk 8]
```diff
@@ -72,15 +73,46 @@ def pdf(self, data):
         data : array-like
             Quantiles, with the last axis of `data` denoting the features.
 
+		weight :  bool, default=False
```
Comment (paradoxysm) [inline dsigm/_core.py:76]: Check indenting

File: dsigm/_core.py [hunk 9]
```diff
@@ -72,15 +73,46 @@ def pdf(self, data):
...
+			Calculate the pdf with the delta of the Core.
+
         Returns
         -------
         pdf : ndarray or scalar
-            Probability density function evaluated at `datas`
+            Probability density function evaluated at `data`.
 		"""
```
Comment (paradoxysm) [inline dsigm/_core.py:83]: Check indenting

File: dsigm/_core.py [hunk 10]
```diff
@@ -72,15 +73,46 @@ def pdf(self, data):
...
+			return p * self.delta
+		else:
+			return p
+
+	def logpdf(self, data, weight=False):
+		"""
+		Log of multivariate normal probability density function.
+
+		Parameters
+        ----------
```
Comment (paradoxysm) [inline dsigm/_core.py]: Check indenting

File: dsigm/mixture/_gmm.py [hunk 11]
```diff
@@ -448,56 +411,53 @@ def bic(self, data):
...
+			The lower the better.
 		"""
-		score = np.exp(self.score(p))
-		fit_coeff = (len(self.cores) * (1 - score)) / (score + 1e-8)
-		fit_gradient = -2 * (90 / (len(self.cores) * (len(self.cores) + fit_coeff))) * len(data)
-		penalty_gradient = (np.square(self.dim) + 1.5 * self.dim + 1) * np.log(len(data))
-		gradient = fit_gradient + penalty_gradient
-		return gradient
+		return np.average([self.bic(data), self.aic(data)],
+							weights=[bic_weight, 1-bic_weight])
```
Comment (paradoxysm) [inline dsigm/mixture/_gmm.py:460]: Check indenting

File: dsigm/mixture/_sgmm.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+from . import GMM
+
+class SGMM(GMM):
+	"""
+	A modified Gaussian Mixture Model that can stabilize
+	the optimal number of components during fitting.
+
+	SGMM refines the number of components during each
+	iteration of the EM algorithm using a narrowing
+	interval based on the Bayesian Information Criterion.
```
Comment (paradoxysm) [inline dsigm/mixture/_sgmm.py]: No longer uses BIC, instead composite ABIC is used

File: dsigm/mixture/_sgmm.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+	Parameters
+	----------
+	init_cores : int, default=5
+		The initial number of Cores (Gaussian components) to fit the data.
+
+	init : {'random', 'kmeans'}, default='kmeans'
+		The method used to initialize the weights, the means and the
+        precisions.
```
Comment (paradoxysm) [inline dsigm/mixture/_sgmm.py]: Check indenting

File: dsigm/mixture/_sgmm.py [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+		Also acts to enable stabilization. If None, stabilization
+		is disabled.
+
+	n_stabilize : int, default=5
+		Number of times the SGMM will run individual fittings during
+		the stabilization process.
+
+	n_init : int, default=10
+		Number of times the SGMM  will be run with different
+        Core seeds. The final results will be the best output of
```
Comment (paradoxysm) [inline dsigm/mixture/_sgmm.py]: Check indenting

File: dsigm/mixture/_sgmm.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+			Interpreted as [min, max].
+
+		abic : tuple, shape (2,)
+			The abic scores corresponding to the interval.
+
+		Returns
+		-------
+		interval : tuple, shape (2,)
+			The interval which contains the optimal number of Cores.
+			Interpreted as [min, max).
```
Comment (paradoxysm) [inline dsigm/mixture/_sgmm.py]: Closed interval

**Review by paradoxysm — COMMENTED** [review]
Code and Documentation check; please see comments.
Reviews on `.ipynb` files still needed

**Review by paradoxysm — COMMENTED** [review]
Approved!
